vespa/methods/rcnn/model.py

```diff
@@ -0,0 +1,385 @@
+            in_features, num_classes
+        )
+
+        # Configure optimizer
+        self.optimizer_name = optimizer_name
+        self.lr = lr
+        self.weight_decay = weight_decay
+        self.optimizer = self.configure_optimizer()
+        self.num_classes = num_classes
+
+    def forward(
+        self,
+        images: List[torch.Tensor],
+        targets: Optional[List[Dict[str, torch.Tensor]]] = None,
+    ):  # noqa
+        """
+        Forward pass for the RCNN model.
+
+        Args:
+            images (List[torch.Tensor]):
+                List of input images as tensors.
+            targets (List[Dict[str, torch.Tensor]], optional):
+                Target annotations for training.
+
+        Returns:
+            If training, returns a dict of losses. Otherwise,
+            returns detections.
+        """
+        return self.model(images, targets)
+
+    def train_model(  # noqa
+        self,
+        train_dataset,
+        batch_size=0,
+        epochs=10,
+        device: str = 'cuda',
+        grad_clip: Optional[float] = None,
+    ):
+        """
+        Train the RCNN model.
+
+        Args:
+            train_dataset: Training dataset.
+            val_dataset: Validation dataset (optional).
+            batch_size (int): Batch size. Defaults to 4.
+            epochs (int): Number of epochs. Defaults to 10.
+            device (str): Device to train on ("cuda" or "cpu").
+                            Defaults to "cuda".
+            grad_clip (float, optional): Max gradient norm for
+                            gradient clipping. Defaults to None.
+        """
+        self.model.to(device)
+
+        train_loader = DataLoader(
+            train_dataset,
+            batch_size=batch_size,
+            shuffle=True,
+            collate_fn=self.custom_collate_fn,
+            num_workers=4,
+            pin_memory=True,
+        )
+
+        for epoch in range(epochs):
+            self.model.train()
+            epoch_loss = 0.0
+
+            for images, targets in train_loader:
+                images = [img.to(device) for img in images]  # noqa
+                targets = [  # noqa
+                    {k: v.to(device) for k, v in t.items()} for t in targets
+                ]  # noqa
+
+                loss_dict = self.model(images, targets)
+                losses = sum(loss for loss in loss_dict.values())
+
+                self.optimizer.zero_grad()
+                losses.backward()
+
+                if grad_clip:
+                    torch.nn.utils.clip_grad_norm_(
+                        self.model.parameters(), grad_clip
+                    )
+
+                self.optimizer.step()
+
+                epoch_loss += losses.item()
+
+            avg_loss = epoch_loss / len(train_loader)
+            print(
+                f'Epoch [{epoch + 1}/{epochs}] - Average Loss: {avg_loss:.4f}'
+            )
+
+    def configure_optimizer(self):
```

Comment:
Se nao for criar uma classe pra essa herdar, esse metodo da pra ficar num utils dos metodos, pq todo mundo usa